walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,112 @@
+        if (event.getAttendees().contains(attendee)) {
+          optionalEvents.add(event);
+          // Prevents duplicate event entry if multiple optional attendees attending same meeting.
+          break;
+        }
+      }
+    }
+
+    // Creates a list of events that optional and required guests are attending, no duplicates.
+    Set<Event> setAllEvents = new HashSet<Event>();
+    setAllEvents.addAll(requiredEvents);
+    setAllEvents.addAll(optionalEvents);
+    ArrayList<Event> allEvents = new ArrayList<>(setAllEvents);
+
+    // Finds all available times that optional and required guests can attend, and returns this
+    //     when there are available times.
+    Collection<TimeRange> allAttendeeTimes = getAvailableTimes(allEvents, request.getDuration());
+    if(allAttendeeTimes.size() > 0) {
+      return allAttendeeTimes;
+    }
+
+    // Returns no times when there are no available times with optional attendees and there are no
+    //     required attendees.
+    if(request.getAttendees().size() == 0){
+      return Arrays.asList(); 
+    }
+
+    // Finds and returns available times for only required attendees(when optional attendees cannot
+    //     attend.
+    return getAvailableTimes(requiredEvents, request.getDuration());
+  }
+
+  /** Returns TimeRange of event duration. */
+  private TimeRange getTime(ArrayList<Event> events, int index) {
+    return events.get(index).getWhen();
+  }
+
+  /** Returns available times given list of existing meetings and desired meeting duration. */
+  private Collection<TimeRange> getAvailableTimes(ArrayList<Event> events, long meetingDuration) {
+    Collections.sort(events, Event.ORDER_BY_START);
+    Collection<TimeRange> possibleTimes = new ArrayList();
+
+    // Returns entire day as available if the requested participants have no existing meetings.
+    if (events.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    for (int i = 0; i <= events.size(); i++) {
+      TimeRange availableTime;
+
+      if (i == 0) {
+        availableTime = TimeRange.fromStartEnd(TimeRange.START_OF_DAY,
+            getTime(events, 0).start(), false);
+      } else if (i == events.size()){
+        availableTime = TimeRange.fromStartEnd(getTime(events, i - 1).end(),
+            TimeRange.END_OF_DAY, true);
+      } else {
+        availableTime = TimeRange.fromStartEnd(getTime(events, i - 1).end(), 
+            getTime(events, i).start(), false);
+      }
+
+      // Adds current time window to possible times if duration is long enough for event.
+      if (availableTime.duration() >= meetingDuration) {
+        possibleTimes.add(availableTime);
+      }
+
+      // Removes next event(s) if it is entirely contained within current event.
+      while(i+1 < events.size()) {
+        if(getTime(events, i).contains(getTime(events, i + 1))) {
+          events.remove(events.get(i + 1));
+        } else {
+          break;
+        }
+      }
```

Comment:
How do you determine which event is more important? What if that second one were?

Reply:
This is only handling the case for when one event is completely contained within another event. So only the meeting which starts first(the events are sorted by start time, so this is the first event), could contain the later event. We only need to know when meetings are, not which meeting is when, so the contained meeting is irrelevant.